lib/web_ui/lib/src/engine/platform_dispatcher/app_lifecycle_state.dart

```diff
@@ -0,0 +1,61 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+import 'package:ui/ui.dart' as ui;
+
+import '../../engine.dart';
+
+/// Signature of functions added as a listener to [ui.AppLifecycleState] changes
+typedef AppLifecycleStateListener = void Function(ui.AppLifecycleState state);
+
+/// Determines the [ui.AppLifecycleState].
+abstract class AppLifecycleStateSupport {
```

Comment:
If this is designed to be an app-level instance (as opposed to view-level), then let's make it a clear singleton by following the same pattern we use for other singletons. The listener callback doesn't have to be provided to the constructor, it can instead be added later. Here's a rough idea:
```dart
abstract class AppLifecycleState {
  static final AppLifecycleState instance = _BrowserAppLifecycleState();

  ui.AppLifecycleState get appLifecycleState => _appLifecycleState;
  ui.AppLifecycleState _appLifecycleState = ui.AppLifecycleState.resumed;

  final List<AppLifecycleStateListener> _listeners = <AppLifecycleStateListener>[];

  void addListener(AppLifecycleStateListener listener) {
    if (_listeners.isEmpty) {
      activate();
    }
    _listeners.add(listener);
    listener(_appLifecycleState);
  }

  void removeListener(AppLifecycleStateListener listener) {
    _listeners.remove(listener);
    if (_listeners.isEmpty) {
      deactivate();
    }
  }

  @protected
  void activate();

  @protected
  void deactivate();

  @visibleForTesting
  void onAppLifecycleStateChange(ui.AppLifecycleState newState) {
    if (newState != _appLifecycleState) {
      _appLifecycleState = newState;
      for (final AppLifecycleStateListener listener in _listeners) {
        listener(newState);
      }
    }
  }
}

class _BrowserAppLifecycleState extends AppLifecycleState {
  @override
  void activate() {
    domWindow.addEventListener('beforeunload', _beforeUnloadListener);
    domDocument.addEventListener('visibilitychange', _visibilityChangeListener);
  }

  @override
  void deactivate() {
    domWindow.removeEventListener('beforeunload', _beforeUnloadListener);
    domDocument.removeEventListener('visibilitychange', _visibilityChangeListener);
  }

  // ...
}
```

Reply:
I'm hesitant to implement the singleton version as it may compromise the existing multi-view support.
https://github.com/flutter/engine/pull/44720#discussion_r1312215165

Reply:
@maRci002 since this is using the visibility observer (which is a global browser attribute), this definitely can be a singleton.

In the future we could receive more signals that may depend on multi-view basis, for example, if we were to use an IntersectionObserver to decide if *any* flutter view is visible in the viewport of the browser or not to mark the app as "invisible"; but for these global browser properties, a singleton is fine!

PS: **Thanks for caring** about the multi-view migration, btw!!